Wpb 22590 1 wiab staging (#101)
wireapp/wire-docs: src/how-to/install/wiab-staging.md

# WIAB Staging Deployment Guide

## Introduction

**Wire in a Box (WIAB) Staging** is a demo installation of Wire running on a single physical machine using KVM-based virtual machines. This setup replicates the multi-node production Wire architecture in a consolidated environment suitable for testing, evaluation, and learning about Wire's infrastructure—but **not for production use**.

> 

Reviewer: As a reader i read it kind of. `similar to production` but it does not establish the core operator intent:
staging should mirror production as closely as practical, within the limits of a single-host KVM setup. 
It sounds like another demo installation where it is s staging guideline.

Author: We can remove the demo word from the definition and replace it with staging.

> **Wire in a Box (WIAB) Staging** is a staging installation of Wire running on a single physical machine using KVM-based virtual machines. This setup replicates the multi-node production Wire architecture in a consolidated environment suitable for testing, evaluation, and learning about Wire's infrastructure—but **not for production use**.

Reviewer 2: ```suggestion
**Wire in a Box (WIAB) Staging** is an installation of Wire running on a single physical machine using KVM-based virtual machines. This setup replicates the multi-node production Wire architecture in a consolidated environment suitable for testing, evaluation, and learning about Wire's infrastructure—but **not for production use**. The main use of this package is to verify that automation inside and outside of the wire product functions in the fashion you expect, before you run said automation in production. This will not test your network environment, load based behaviors, or the interface between wire and it's callingservices when using a DMZ'd network configuration.
```

> **Important:** This is a sandbox environment. Data from a staging installation **cannot be migrated to production**. WIAB Staging is designed for experimentation, validation, and understanding Wire's deployment model.

> 

Reviewer: This is obvious, isn't it? It seems we are telling users in multiple places, `don't do this`. If we are not certain that the user understand what is staging and prod env means, I would suggest to create a reference for these terms.

Author: This note is to remind that staging to production migration isn't possible and a user should start with a staging setup and then plan for production setup but the initial staging can't be converted to production. 

We have created a section below ## Relation to production to make it more clear.

For a high-level comparison of WIAB Dev, WIAB Staging, and Production, see the [planning overview](planning.md) (in particular, [Artifact bundle and offline deployment](planning.md#artifact-bundle-and-offline-deployment)).

## Architecture overview

![Wire in a Box Staging Architecture](img/architecture-wiab-stag.png)

## Relation to production

WIAB Staging is designed to be structurally similar to a production deployment:

- Separate nodes for Kubernetes and stateful services.
- Use of offline artifacts, Ansible, and Helm for installation similar to production workflows.
- Able to test upgrade proceedures for production.

Key differences from production:

- All VMs run on a single physical host (single failure domain).
- Resource sizing is optimized for staging, not for peak production traffic.
- Calling services run with limited resources (replicas=1).

> - [Calling services](../../understand/overview.md#calling) will share the same k8s cluster as Wire stateful services hence, all infrastructure will be DMZ (De-militarized zone).

Reviewer 2: I'm not sure how the DMZ concept applies here.

Author: It is becoming DMZ as all the calling traffic (udp port 32768-65535) is being allowed on the same adminhost firewall which is serving the HTTPS traffic for other wire services. Calling pods are in the same k8s cluster as wire-server pods

Reviewer 2: yeah, DMZ as a concept is about separation. this is not separated, so it doesn't fit.

Author: yes, its better to call it open - let me rephrase it

If you need a fully supported, highly-available, multi-datacenter deployment, use the **Production** path instead (see `ansible-VMs.md` and `helm-prod.md`).

> 

Reviewer: What if the suggested staging env (hardware/network infra etc) is enough for a user to run their prod setup?

Author: As per the definition of production a staging doesn't satisfy the condition, Hence we are specifying why a staging is not a production:
- no HA, single failure domain
- no network security against wire-components 
- limited wire-calling capacity 

A user should never think that a staging is production.

Reviewer: `A user should never think that a staging is production.` agreed but my question was if that is enough for the user to run their production. I understand what you mean here. Then to be fair, I would argue to have the same setup for staging as like prod.

Author: A staging solution is not production due to no HA, no equivalent security as we achieve in prod and limited calling scaling. However, our wiab-staging solution is a resource efficient way to test the same tools we have for prod, to test upgrades or security fixes before implementing them in production. If staging is exactly same as prod then this would cost a huge amount to customer to be able to manage a staging solution especially when there is no traffic on the staging solution. If there are some cases which our staging solution isn't sufficient then we can work towards improving it.

Reviewer: isn't it up to the customer/user how would they like to setup their staging environment? When we suggest a staging environment why that will differ too much from production other than a lower capacity as it does not need to scale based on higher traffic?  We should then call it a test environment. 
As example of how staging env should be defined: https://northflank.com/blog/what-is-a-staging-environment-how-to-set-one-up

Author: Yes, a user can choose how they want to set up their staging environment. For example, they can use a cloud provider and skip all Ansible-based service provisioning.

This guide is intended for users who are not familiar with Wire but want to understand how a Wire installation works, test upgrades, and have a minimal production-like Wire infrastructure.

This solution has been developed internally at Wire over the past few years and is intended to function as a staging environment.

We can discuss the naming and history of this solution internally in a separate ticket.

## Requirements

- One physical machine with hypervisor support (KVM):
  - **Memory:** 55 GiB RAM
  - **Compute:** 29 vCPUs  (At least processor 6 cores)
  - **Storage:** 850 GB disk space (thin-provisioned)

>   - 7 VMs with [Ubuntu 22](https://releases.ubuntu.com/jammy/) as per (#VM-Provisioning)

Reviewer 2: how is this a requirement? we give these.

Author: yes, its optional, let me make it clear

- **DNS Records**: 

>     - a way to create DNS records for your domain name (e.g. wire.example.com) 

Reviewer 2: ```suggestion
    - A method to create DNS records for your domain name (e.g. wire.example.com) 
```

    - See [DNS Requirements for Wire Deployments](dns-requirements.md) for the full list of required hostnames and examples.
- **SSL/TLS certificates**:

>     - a way to create SSL/TLS certificates for your domain name (to allow connecting via https://)

Reviewer 2: ```suggestion
    - A method to create SSL/TLS certificates for your domain name (to allow connecting via https://)
```

    - To simplify certificate management, we recommend using **Let’s Encrypt** together with **cert-manager** where internet access is available. See [TLS and Certificates](tls-certificates.md) to choose between cert‑manager + Let’s Encrypt and bring‑your‑own certificates.
- **Network**: 

>     - SSH to the physical host and to the VMs.

Reviewer 2: ```suggestion
    - SSH access to the physical host.
```

    - HTTP/HTTPS and UDP traffic from clients to the cluster, without interference from host-level firewalls (UFW, nftables) on required ports.
    - IP forwarding between the host’s network interfaces when following the reference topology. Check [Network Traffic Configuration](#network-traffic-configuration) for more details.
- **Wire-server-deploy artifact**: Access to a `wire-server-deploy` offline artifact bundle (contact Wire support for the latest stable artifact). This tarball contains all required Bash scripts, deb packages, Ansible playbooks, Helm charts and container images to perform the installation. See the [planning overview](planning.md#artifact-bundle-and-offline-deployment) for how artifacts are used across WIAB Dev, WIAB Staging and Production.

> > Note: Check [WIAB Dev](./planning.md#wiab-dev-single-vm-wire-in-a-box) solution if looking for more light weight installation of wire-server.

Reviewer 2: ```suggestion
> Note: Check [WIAB Dev](./planning.md#wiab-dev-single-vm-wire-in-a-box) solution if looking for a lighter weight installation of wire-server, for development uses.
```

## VM Provisioning

> We would require 7 VMs as per the following details, you can choose to use your own hypervisor to manage the VMs or use our [Wiab staging ansible provisioning playbook](https://github.com/wireapp/wire-server-deploy/blob/master/ansible/wiab-staging-provision.yml) against your physical node to setup the VMs.

Reviewer 2: ```suggestion
Our deployment will be into 7 VMs, shown in the below VM Archetecture and Resource Allocation table, You can choose to use your own hypervisor to manage the VMs or use our [Wiab staging ansible provisioning playbook](https://github.com/wireapp/wire-server-deploy/blob/master/ansible/wiab-staging-provision.yml) against your physical node to setup the VMs.
```

**VM Architecture and Resource Allocation:**

| Hostname | Role | RAM | vCPUs | Disk |
|----------|------|-----|-------|------|
| assethost | Asset/Storage Server | 4 GiB | 2 | 100 GB |
| kubenode1 | Kubernetes Node 1 | 9 GiB | 5 | 150 GB |
| kubenode2 | Kubernetes Node 2 | 9 GiB | 5 | 150 GB |
| kubenode3 | Kubernetes Node 3 | 9 GiB | 5 | 150 GB |
| datanode1 | Data Node 1 | 8 GiB | 4 | 100 GB |
| datanode2 | Data Node 2 | 8 GiB | 4 | 100 GB |
| datanode3 | Data Node 3 | 8 GiB | 4 | 100 GB |
| **Total** | | **55 GiB** | **29** | **850 GB** |

*Note: These specifications are optimized for testing and validation purposes, not for performance benchmarking.*

**VM Service Distribution:**

- **kubenodes (kubenode1, kubenode2, kubenode3):** Run the Kubernetes cluster and host Wire backend services
- **datanodes (datanode1, datanode2, datanode3):** Run distributed data services:
  - Cassandra (distributed database)

>   - PostgreSQL (operational database)

Reviewer: What is operation database and distributed database here?

Author: The intent was to mention distributed (all nodes equal) and operational (Primary RW and secondary RO) nodes. I will remove it to avoud confusion

  - Elasticsearch (search engine)
  - Minio (S3-compatible object storage)
  - RabbitMQ (message broker)
- **assethost:** Hosts static assets like container images and debian binaries to be used by kubenodes and datanodes during installation.

## Getting the Ansible playbooks

> **Note:** If you prefer to manage VMs yourself with another hypervisor, you can skip this step and instead create the VMs manually with the specs shown above. Along with VM creation, you need to perform a few more steps like downloading the [wire-server-deploy artifact](planning.md#artifact-bundle-and-offline-deployment) and [setup network configuration](#network-traffic-configuration) including the firewall rules.

> 

Reviewer: after downloading the artifact, the user can run [offline-vm-setup.sh](https://github.com/wireapp/wire-server-deploy/blob/master/bin/offline-vm-setup.sh) too.

Why we would suggest a user to clone the repo when we have delivered the artifacts and capable of running the deployment.

Its not a contribution guidelines usually where it as asked to clone/fork etc.

Author: Clone or fork are not necessary, we are sharing the guide to download the ansible playbooks via wget (https://github.com/wireapp/wire-docs/pull/101/changes#diff-ffa83634d14fb8bd04fd337b1199e4d77d12e05ead894298c030d349987a6edbR92) to whole repository. 

There are 2 ways of installation possible here:
1. A user choose to use our Ansible playbooks to setup VMs (and required packages before hand) and also download the artifact, then create an inventory based on the VMs created above.
2. A user creates their own VMs using their own hypervisor, they can manually download the artifact, manually edit the inventory file and then continue with the next steps. 

The first step requires user to download the ansible playbooks before hand. 

A user can run the script https://github.com/wireapp/wire-server-deploy/blob/master/bin/offline-vm-setup.sh directly but the user would still be required to install all the required deb packages on the machine to enable the VM creation. Then user would be required to edit the inventory manually. At the end user would be required to setup the firewall (nftables) rules manually if not using the ansible playbook.

Reviewer: Yes, and all the deployment supported scripts/playbooks are in the artifacts. Just you know I am strongly against suggesting the user to clone/download the whole repo to setup their environment.

Author: I am not sure what the best approach is for sharing an Ansible-based solution with users to help them get started.

In the case of wiab-staging, there are two layers of Ansible involved:

Adminhost provisioning layer
This includes downloading the necessary tools to provision VMs, fetching the artifacts, and setting up the firewall on the adminhost. All these operations are performed directly on the adminhost.

VM configuration layer
This involves configuring the VMs, which can either be created using our Ansible playbooks or via another hypervisor solution. These operations are executed from the wire-server-deploy-container (WSD container) by mounting the entire wire-server-deploy directory. This setup is expected to provide the required SSH keys or agent access for configuring the VMs.

We are currently not interacting with the adminhost from within the WSD container. If we were to do so, we would need to ask users to provide SSH access inside the container, which goes against the separation between provisioning the staging infrastructure and configuring the VMs.

Most Ansible-based solutions are distributed either as collections or assume that users already have playbooks to execute them, as seen in Kubespray:
https://github.com/kubernetes-sigs/kubespray/blob/master/docs/ansible/ansible_collection.md

Sharing a tarball artifact of Ansible playbooks for provisioning nodes is quite uncommon.

Please let me know if you have any better suggestions for handling the provisioning in this setup.

> Note: If you have already used the [Wiab staging ansible provisioning playbook](#provisioning-the-vms-with-the-wiab-staging-playbook) to set up VMs, this file should have been prepared for you.

In that inventory you will:

- Set `ansible_user` (this user should have passwordless `sudo` access) and `ansible_ssh_private_key_file` under `all` vars.
- Assign each VM to the appropriate groups, for example:
  - `kube-master`, `kube-node`, `etcd` for Kubernetes nodes.
  - `datanodes` which will further map to `cassandra`, `cassandra_seed`, `elasticsearch`, `minio`, `rmq-cluster`, `postgresql` for data services.
  - `assethost` for assethost service

For a deeper explanation of offline inventories and VM roles in production-like setups, see [Installing kubernetes and databases on VMs with ansible](ansible-VMs.md#ansible-vms).

## Next steps

> Since the secondary inventory is ready, please continue with the following steps:

Reviewer 2: ```suggestion
Once the secondary inventory is ready, please continue with the following steps:
```

  ```bash
  # without the alias `d`
  ./bin/offline-secrets.sh
  ```
  Run to generate fresh secrets for Minio, coturn and Wire services. This creates secret files such as `ansible/inventory/group_vars/all/secrets.yaml` and `values/wire-server/secrets.yaml`. To understand which secrets are generated and where they are stored across Ansible and Helm, see [Secrets Overview](secrets-overview.md).

### Kubernetes & Data Services Deployment

- **Deploying Kubernetes and stateful services**
  ```bash
  d ./bin/offline-cluster.sh
  ```
  Run to deploy Kubernetes and stateful services (Cassandra, PostgreSQL, Elasticsearch, Minio, RabbitMQ). This script deploys all datastore infrastructure needed for Wire backend operations.

### Helm Operations to install Wire services and supporting Helm charts

**Helm chart deployment (automated):** The script `bin/helm-operations.sh` will deploy the charts for you. It prepares `values.yaml`/`secrets.yaml`, customizes them for your domain/IPs, then runs Helm installs in the correct order. For a detailed, production-focused reference of the same charts, see [Installing wire-server (production) components using Helm](helm-prod.md#helm-prod).

> **User-provided inputs (set these before running):**

Reviewer 2: set these before running.. what? the offline-helm we told them to run above?

Author: we didn't run any helm operations above. We mentioned which script will deploy helm operations and what else that script will do. 
https://github.com/wireapp/wire-docs/pull/101/changes#diff-ffa83634d14fb8bd04fd337b1199e4d77d12e05ead894298c030d349987a6edbR207

- If you **do not** want Let’s Encrypt / cert-manager (for example, you are using **[Bring Your Own certificates](docs_ubuntu_22.04.md#acquiring--deploying-ssl-certificates)**), disable this step by passing env variable `DEPLOY_CERT_MANAGER=FALSE` when running `bin/helm-operations.sh`.
  - When choosing `DEPLOY_CERT_MANAGER=FALSE`, ensure your ingress is configured and deployed with your own TLS secret(s) as described at [TLS and Certificates](tls-certificates.md).
  - When choosing `DEPLOY_CERT_MANAGER=TRUE`, ensure if further network configuration is required by following [cert-manager behaviour in NAT / bridge environments](#cert-manager-behaviour-in-nat--bridge-environments).

**To run the automated helm chart deployment with your variables**:
```bash
# example command - verify the variables before running it
d sh -c 'TARGET_SYSTEM="example.dev" CERT_MASTER_EMAIL="certmaster@example.dev" DEPLOY_CERT_MANAGER=TRUE ./bin/helm-operations.sh'
```

**Charts deployed by the script:**
- External datastores and helpers: `cassandra-external`, `elasticsearch-external`, `minio-external`, `rabbitmq-external`, `postgresql-external`, `databases-ephemeral`, `reaper`, `fake-aws`, `demo-smtp`.

> - Wire services: `wire-server`, `webapp`, `account-pages`, `team-settings`, `smallstep-accomp`.

Reviewer 2: why do we have smallstep anything here?

Author: It has been recently removed, that PR has not merged, will make this change.

- Defining your network variables (Coturn IP, Kubernetes node IP, WAN interface)
- Creating NAT rules for HTTP/HTTPS ingress traffic
- Setting up TURN protocol forwarding for Coturn and traffic for SFTD

*Note: If you have already ran the playbook wiab-staging-provision.yml then it is already be configured for you. Confirm it by checking if the wire endpoint `https://webapp.TARGET_SYSTEM` is reachable from public internet or your private network (in case of private network), but not from the adminhost itself.*

You can also apply these rules using the Ansible playbook against your adminhost, by following:

```bash
# create the inventory.yml before running it
ansible-playbook -i inventory.yml ansible/wiab-staging-nftables.yml
```

You can run the above playbook from local system or where you have cloned/downloaded the [Wire server deploy ansible playbooks](#getting-the-ansible-playbooks).

The inventory `inventory.yml` should define the following variables:

```ini
[all:vars]
# Kubernetes node IPs
kubenode1_ip=192.168.122.11
kubenode2_ip=192.168.122.12
kubenode3_ip=192.168.122.13

> # Calling services node (usually kubenode3)

Reviewer 2: 'usually'?

Author: This will be updated post merging https://github.com/wireapp/wire-server-deploy/pull/881 
This is in review - they are related
